serverless/experimental/gcp-log/function.go

```diff
@@ -12,27 +12,149 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package p provides Google Cloud Function for sequencing entries in a
-// serverless log.
+// Package p provides Google Cloud Functions for adding (sequencing and
+// integrating) new entries to a serverless log.
 package p
 
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"net/http"
 	"os"
 
 	"github.com/transparency-dev/merkle/rfc6962"
 	"golang.org/x/mod/sumdb/note"
+	"google.golang.org/api/iterator"
 
 	"github.com/google/trillian-examples/serverless/pkg/log"
 	"github.com/gcp_serverless_module/internal/storage"
 
 	fmtlog "github.com/google/trillian-examples/formats/log"
-	golog "log"
 )
 
+func validateCommonArgs(w http.ResponseWriter, origin string) (ok bool, pubKey string) {
+	// TODO(jayhou): do we need this?
+	if len(origin) == 0 {
+		http.Error(w, "Please set `origin` in HTTP body to log identifier.", http.StatusBadRequest)
+		return false, ""
+	}
+
+	pubKey = os.Getenv("SERVERLESS_LOG_PUBLIC_KEY")
+	if len(pubKey) == 0 {
+		http.Error(w,
+			"Please set SERVERLESS_LOG_PUBLIC_KEY environment variable",
+			http.StatusBadRequest)
+		return false, ""
+	}
+
+	return true, pubKey
+}
+
+// Sequence is the entrypoint of the `sequence` GCF function.
+func Sequence(w http.ResponseWriter, r *http.Request) {
+	// TODO(jayhou): validate that EntriesDir is only touching the log path.
+
+	var d struct {
+		Bucket string `json:"bucket"`
+		EntriesDir string   `json:"entriesDir"`
+		Origin     string `json:"origin"`
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&d); err != nil {
+		fmt.Printf("json.NewDecoder: %v", err)
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+
+	ok, pubKey := validateCommonArgs(w, d.Origin)
+	if !ok {
+		return
+	}
+
+	// init storage
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, os.Getenv("GCP_PROJECT"), d.Bucket)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to create GCS client: %q", err), http.StatusBadRequest)
+		return
+	}
+
+	// Read the current log checkpoint to retrieve next sequence number.
+
+	cpRaw, err := client.ReadCheckpoint(ctx)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to read log checkpoint: %q", err), http.StatusBadRequest)
+		return
+	}
+
+	// Check signatures
+	v, err := note.NewVerifier(pubKey)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to instantiate Verifier: %q", err), http.StatusBadRequest)
+		return
+	}
+	cp, _, _, err := fmtlog.ParseCheckpoint(cpRaw, d.Origin, v)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to parse Checkpoint: %q", err), http.StatusBadRequest)
+		return
+	}
+	client.SetNextSeq(cp.Size)
+
+	// sequence entries
+
+	h := rfc6962.DefaultHasher
+	it := client.GetObjects(ctx, d.EntriesDir)
```

Comment:
Does anything delete pending entries once they're successfully sequenced, or is that for a later PR?

Reply:
Are you referring to the directory `<rootDir>/leaves/pending/...`? I know in the Github Actions implementation, we created this directory [*] in order to write a temporary file which we remove after the full sequencing. I think we discussed that GCS does not have the concept of linking, so for the GCP implementation, I directly write the file without writing a temporary file first.

[*] https://github.com/google/trillian-examples/blob/ca80f135b82b9173e1bfbf45feec841edf01b5bf/serverless/internal/storage/fs/fs.go#L133

Reply:
Ah, I was thinking of this: 

https://github.com/google/trillian-examples/blob/ca80f135b82b9173e1bfbf45feec841edf01b5bf/serverless/deploy/github/log/sequence_and_integrate/entrypoint.sh#L37

We should probably pop a pin in to think about this later - it might be better for the `sequence` tool to (optionally?) remove source leaf files one they've been sequenced since it knows for sure which ones have been.

wdyt?